### PR TITLE
Use flyway to perform database migrations on startup

### PR DESCRIPTION
**Pull request by posulliv**
## Description

Add support for running mirations automatically with flyway on server startup. Since users may not want this, I added a config option to opt out of having the gateway process perform migrations. Migrations will only be performed by the gateway process if this config option is enabled.

## Additional context and related issues

This change will allow us to add backend DB support for Oracle in the future.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( X) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by mosabua** [comment]
Awesome.. we just talked about this recently in the dev sync - fyi @willmostly

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java [hunk 1]
```diff
@@ -20,14 +20,16 @@ public class DataStoreConfiguration
     private String password;
     private String driver;
     private Integer queryHistoryHoursRetention = 4;
+    private boolean runMigrationsEnabled;
```
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java:23]: This could be defaulted to `true`. Most users should allow the gateway to manage their DB schema, and it makes things easier for new users as well.

File: gateway-ha/src/test/java/io/trino/gateway/ha/persistence/BaseTestDatabaseMigrations.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    protected abstract String getDriver();
+
+    @AfterAll
+    public final void close()
+    {
+        container.close();
+    }
+
+    @Test
+    public void testMigrationWithEmptyDatabase()
```
Comment (willmostly) [inline gateway-ha/src/test/java/io/trino/gateway/ha/persistence/BaseTestDatabaseMigrations.java:51]: Can we add a test case for a non-empty database as well?

File: gateway-ha/src/test/java/io/trino/gateway/ha/persistence/BaseTestDatabaseMigrations.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    protected void dropAllTables()
+    {
+        String gatewayBackendTable = "DROP TABLE IF EXISTS gateway_backend";
+        String queryHistoryTable = "DROP TABLE IF EXISTS query_history";
+        String propertiesTable = "DROP TABLE IF EXISTS resource_groups_global_properties";
+        String resourceGroupsTable = "DROP TABLE IF EXISTS resource_groups";
+        String selectorsTable = "DROP TABLE IF EXISTS selectors";
+        String exactMatchTable = "DROP TABLE IF EXISTS exact_match_source_selectors";
+        String flywayHistoryTable = "DROP TABLE IF EXISTS flyway_schema_history";
+        Handle jdbiHandle = jdbi.open();
```
Comment (willmostly) [inline gateway-ha/src/test/java/io/trino/gateway/ha/persistence/BaseTestDatabaseMigrations.java:92]: Can we add a check that the schema is now empty? A query against `information_schema` should work for both pg and mysql. This will ensure that if someone adds a table to the schema, they remember to update this method

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java (hunk 1 above)
Comment (willmostly) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java:23]: Please document this setting

**Review by willmostly — COMMENTED** [review]
Just a few minor comments - this looks good to me overall and should be a nice improvement!

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java (hunk 1 above)
Comment (willmostly) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java]: I think https://github.com/trinodb/trino-gateway/blob/f4b78f48ca5e882575a4f82640bb70940195c79b/docs/installation.md?plain=1#L39 is probably the best place for it now. You can revise the section that refers to initializing the DB manually and let users know that the Gateway will do it for them, unless they set `runMigrationsEnabled: false`, in which case they will need to manually run the schema SQLs

**Comment by mosabua** [comment]
Don't have time to review but here are some thoughts:

- Can we check database permissions before scripts are run
- I agree that we want to have this migration on by default, but be able to deactivate it
- Must be part of the app start so it works for any deployment
- If migration fails or some problem is found the app startup should probably fail hard so the app is not used at all in some weird state
- have to figure out how this works in a cluster where we have multiple instance, maybe first instance gets a lock and then all other instances are prevented from starting (or fail to start) until the migration is done

**Comment by posulliv** [comment]
thanks @mosabua Replying to your comments:

> Can we check database permissions before scripts are run

flyway will fail if database permissions are incorrect and the gateway process will not start up. trino has the same behavior for database backed resource groups.

> I agree that we want to have this migration on by default, but be able to deactivate it

Got it. This is implemented in the PR now with an option to disable it.

> Must be part of the app start so it works for any deployment

This is in the PR. Migrations occur on startup unless they are explitly disabled in the data store config.

> If migration fails or some problem is found the app startup should probably fail hard so the app is not used at all in some weird state

Yep, this is behavior as implemented in this PR. If a migration fails the process will not start and the logs will contain the migration error.

> have to figure out how this works in a cluster where we have multiple instance,

Are you saying we need to handle this in this PR? This state indicating one gateway process is performed migrations would need to be maintained somewhere. Seems like backing DB is the only only place we could store it when there are multiple gateway processes?

**Comment by mosabua** [comment]
Awesome.. thanks for answering @posulliv 

The only thing I think we need to figure out is how to deal with multiple nodes. Can you find out how flyway deals with this. Feels like it should support that out of the box somehow.

Maybe the first node that starts up locks the db for the upgrade and any additional nodes connecting just fail or have to wait until migration is completed. We just need to avoid having multiple nodes connecting and all of them starting to migrate the database and ending up with a mess

**Comment by posulliv** [comment]
Yeah @willmostly suggested flyway might take care of this as well. It does in fact use locks - https://documentation.red-gate.com/fd/postgresql-database-235241807.html#

> By default Flyway uses a transactional lock with PostgreSQL

So the behavior should be the first gateway instance to start will get the lock and run the migrations. Other instances might fail during startup but if this is a k8s deployment, once they are automatically restarted they should be good as migrations will be completed.

**Comment by mosabua** [comment]
Sweet.. lets just add that to the docs then.

**Comment by posulliv** [comment]
@mosabua ok cool. I updated the docs in latest push.

File: docs/installation.md [hunk 4]
```diff
@@ -38,16 +38,37 @@ distribution is installed.
...
+dataStore:
+  jdbcUrl: jdbc:postgresql://postgres:5432/trino_gateway_db
+  user: USER
+  password: PASSWORD
+  driver: org.postgresql.Driver
+  queryHistoryHoursRetention: 24
+  runMigrationsEnabled: false
+```
+
+`flyway` uses a transactional lock in databases that support it such as
```
Comment (willmostly) [inline docs/installation.md:64]: Nit: Capitalize  `flyway`

**Review by willmostly — APPROVED** [review]
Looks good!

**Comment by willmostly** [comment]
@mosabua ptal
